File: python/src/eurocreator/sii.py
"""Writer for SII definition files.

SII is SCS' plain-text unit format. The game's parser is unforgiving about a
few things that are easy to get wrong by hand, and this module makes them
impossible: unit names are validated, strings are escaped, arrays use the
``key[]:`` form, and files are written with the exact ``SiiNunit`` wrapper.
"""

from __future__ import annotations

from pathlib import Path
from typing import Iterable, List, Mapping, Sequence, Union

from .naming import is_unit_name

__all__ = ["SiiUnit", "SiiFile", "write_sui"]

Value = Union[str, int, float, bool, Sequence[float]]


def _format_value(value: Value) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        return f"{value:g}"
    if isinstance(value, (tuple, list)):
        return "(" + ", ".join(f"{float(v):g}" for v in value) + ")"
    # Strings: the parser has no escape for a literal quote, so reject rather
    # than emit a file that silently truncates.
    text = str(value)
    if '"' in text:
        raise ValueError(f"SII string values cannot contain a double quote: {text!r}")
    return f'"{text}"'


class SiiUnit:
    """One ``class_name : unit_name { ... }`` block."""

    def __init__(self, class_name: str, unit_name: str):
        self.class_name = class_name
        self.unit_name = unit_name
        self._lines: List[str] = []

    def set(self, key: str, value: Value) -> "SiiUnit":
        self._lines.append(f"\t{key}: {_format_value(value)}")
        return self

    def append(self, key: str, value: Value) -> "SiiUnit":
        """Add one entry to an array attribute, i.e. ``key[]: value``."""
        self._lines.append(f"\t{key}[]: {_format_value(value)}")
        return self

    def extend(self, key: str, values: Iterable[Value]) -> "SiiUnit":
        for value in values:
            self.append(key, value)
        return self

    def include(self, sui_name: str) -> "SiiUnit":
        """Pull in a shared ``.sui`` fragment.

        ``@include`` must sit at column zero -- indenting it makes the parser
        treat the line as an attribute and fail the whole unit.
        """
        self._lines.append(f'@include "{sui_name}"')
        return self

    def blank(self) -> "SiiUnit":
        self._lines.append("")
        return self

    def render(self) -> str:
        head = f"{self.class_name}: {self.unit_name}\n{{\n"
        return head + "\n".join(self._lines) + "\n}\n"


class SiiFile:
    """A ``SiiNunit { ... }`` document holding one or more units."""

    def __init__(self, *units: SiiUnit):
        self.units: List[SiiUnit] = list(units)

    def add(self, unit: SiiUnit) -> SiiUnit:
        self.units.append(unit)
        return unit

    def render(self) -> str:
        body = "\n".join(unit.render() for unit in self.units)
        return "SiiNunit\n{\n" + body + "}\n"

    def write(self, path: Path) -> Path:
        path = Path(path)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(self.render(), encoding="utf-8")
        return path


def write_sui(path: Path, attributes: Mapping[str, Value]) -> Path:
    """Write a ``.sui`` fragment: bare indented attributes, no unit wrapper."""
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    lines = [f"\t{key}: {_format_value(value)}" for key, value in attributes.items()]
    path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    return path

File: packages/core/src/sii.ts
/**
 * Writer for SII definition files.
 *
 * SII is SCS' plain-text unit format. The game's parser is unforgiving about a
 * few things that are easy to get wrong by hand, and this module makes them
 * impossible: strings are checked, arrays use the `key[]:` form, `@include`
 * sits at column zero, and files carry the exact `SiiNunit` wrapper.
 */

export type SiiValue = string | number | boolean | readonly number[];

function formatValue(value: SiiValue): string {
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new Error(`SII cannot represent ${value}`);
    return Number.isInteger(value) ? String(value) : formatFloat(value);
  }
  if (Array.isArray(value)) return `(${value.map((v) => formatFloat(Number(v))).join(", ")})`;
  const text = String(value);
  // The parser has no escape for a literal quote, so reject rather than emit a
  // file that silently truncates at the stray quote.
  if (text.includes('"')) {
    throw new Error(`SII string values cannot contain a double quote: ${JSON.stringify(text)}`);
  }
  return `"${text}"`;
}

/** Match Python's `%g`: shortest round-tripping form, trailing zeros dropped. */
function formatFloat(value: number): string {
  if (Number.isInteger(value)) return String(value);
  const fixed = value.toPrecision(6).replace(/0+$/, "").replace(/\.$/, "");
  return String(Number(fixed));
}

/** One `class_name : unit_name { ... }` block. */
export class SiiUnit {
  private readonly lines: string[] = [];

  constructor(
    readonly className: string,
    readonly unitName: string,
  ) {}

  set(key: string, value: SiiValue): this {
    this.lines.push(`\t${key}: ${formatValue(value)}`);
    return this;
  }

  /** Add one entry to an array attribute, i.e. `key[]: value`. */
  append(key: string, value: SiiValue): this {
    this.lines.push(`\t${key}[]: ${formatValue(value)}`);
    return this;
  }

  extend(key: string, values: Iterable<SiiValue>): this {
    for (const value of values) this.append(key, value);
    return this;
  }

  /**
   * Pull in a shared `.sui` fragment.
   *
   * `@include` must sit at column zero -- indenting it makes the parser treat
   * the line as an attribute and fail the whole unit.
   */
  include(suiName: string): this {
    this.lines.push(`@include "${suiName}"`);
    return this;
  }

  blank(): this {
    this.lines.push("");
    return this;
  }

  render(): string {
    return `${this.className}: ${this.unitName}\n{\n${this.lines.join("\n")}\n}\n`;
  }
}

/** A `SiiNunit { ... }` document holding one or more units. */
export class SiiFile {
  readonly units: SiiUnit[];

  constructor(...units: SiiUnit[]) {
    this.units = units;
  }

  add(unit: SiiUnit): SiiUnit {
    this.units.push(unit);
    return unit;
  }

  render(): string {
    return `SiiNunit\n{\n${this.units.map((u) => u.render()).join("\n")}}\n`;
  }
}

/** Render a `.sui` fragment: bare indented attributes, no unit wrapper. */
export function renderSui(attributes: ReadonlyArray<readonly [string, SiiValue]>): string {
  return attributes.map(([key, value]) => `\t${key}: ${formatValue(value)}`).join("\n") + "\n";
}
